### PR TITLE
Feature/odometry msg

### PR DESCRIPTION
**Pull request by burrimi**

### CODE REVIEW

**Comment by simonlynen** [comment]
add to whitelist

File: msf_core/include/msf_core/implementation/msf_core_inl.h [hunk 1]
```diff
@@ -137,6 +137,8 @@ void MSF_Core<EKFState_T>::ProcessIMU(
   // Get inputs.
   currentState->a_m = linear_acceleration;
   currentState->w_m = angular_velocity;
+  currentState->noise_gyr = Vector3::Constant(usercalc_.GetParamNoiseGyr());
```
Comment (markusachtelik) [inline msf_core/include/msf_core/implementation/msf_core_inl.h]: just for completeness, do we want to have acceleration here as well?

File: msf_core/include/msf_core/implementation/msf_core_inl.h [hunk 2]
```diff
@@ -787,6 +790,8 @@ shared_ptr<EKFState_T> MSF_Core<EKFState_T>::GetClosestState(double tstamp) {
               / (nextState->time - lastState->time)
               * (timenow - lastState->time);
 
+      currentState->noise_gyr = Vector3::Constant(usercalc_.GetParamNoiseGyr());
```
Comment (markusachtelik) [inline msf_core/include/msf_core/implementation/msf_core_inl.h]: good catch!

File: msf_core/include/msf_core/implementation/msf_state_inl.h [hunk 3]
```diff
@@ -344,6 +371,87 @@ void GenericState_T<stateVector_T, StateDefinition_T>::GetPoseCovariance(
...
+      q_type, msf_tmp::CorrectionStateLengthForType>::value;
+
+  /*        |  cov_v_v  |  cov_v_q  |
+   *        |           |           |
+   * cov =  |-----------|-----------|
+   *        |           |           |
+   *        |  cov_q_v  |  cov_q_q  |
+   */
+
+  for (int i = 0; i < 9; i++)
```
Comment (markusachtelik) [inline msf_core/include/msf_core/implementation/msf_state_inl.h]: better do this with Eigen::Map<> and block access?

File: msf_core/include/msf_core/implementation/msf_core_inl.h [hunk 4]
```diff
@@ -236,6 +238,7 @@ void MSF_Core<EKFState_T>::ProcessExternallyPropagatedState(
   // Get inputs.
   currentState->a_m = linear_acceleration;
   currentState->w_m = angular_velocity;
+  currentState->noise_gyr = Vector3::Constant(usercalc_.GetParamNoiseGyr());
```
Comment (simonlynen) [inline msf_core/include/msf_core/implementation/msf_core_inl.h]: same as above

File: msf_core/include/msf_core/implementation/msf_state_inl.h [hunk 5]
```diff
@@ -327,23 +355,83 @@ void GenericState_T<stateVector_T, StateDefinition_T>::GetPoseCovariance(
...
+  cov.block<3,3>(0,3) = P.template block<3,3>(idxstartcorr_v, idxstartcorr_q);
+  cov.block<3,3>(3,3) = P.template block<3,3>(idxstartcorr_q, idxstartcorr_q);
 }
 
+/// Writes the covariance corresponding to velocity and angular velocity expressed in the imu frame
+// to cov.
+template<typename stateVector_T, typename StateDefinition_T>
+void GenericState_T<stateVector_T, StateDefinition_T>::GetTwistCovarianceInImuFrame(
+    geometry_msgs::TwistWithCovariance::_covariance_type& cov) {
+
```
Comment (simonlynen) [inline msf_core/include/msf_core/implementation/msf_state_inl.h]: vsp

File: msf_core/include/msf_core/msf_state.h [hunk 6]
```diff
@@ -142,6 +143,9 @@ struct GenericState_T {
   Eigen::Matrix<double, 3, 1> w_m;         ///< Angular velocity from IMU.
   Eigen::Matrix<double, 3, 1> a_m;         ///< Linear acceleration from IMU.
 
+  // imu uncertainty
```
Comment (simonlynen) [inline msf_core/include/msf_core/msf_state.h]: fix comment

File: msf_core/include/msf_core/implementation/msf_state_inl.h [hunk 7]
```diff
@@ -327,23 +355,83 @@ void GenericState_T<stateVector_T, StateDefinition_T>::GetPoseCovariance(
...
+                    0, 0, noise_gyr[2] * noise_gyr[2];
+
+  CHECK_EQ(cov.size(), 6 * 6);
+  Eigen::Map<Eigen::Matrix<double,6,6> > covariance_map(cov.data());
+
+  covariance_map.block<3,3>(0,0) = cov_velocity_I;
+
+  // Add noise of gyro measurement and gyro bias to get covariance of corrected angular velocity.
+  covariance_map.block<3,3>(3,3) = P.template block<3,3>(idxstartcorr_b_w, idxstartcorr_b_w) +
+                                   cov_noise_gyr;
```
Comment (simonlynen) [inline msf_core/include/msf_core/implementation/msf_state_inl.h]: time scaling

**Comment by burrimi** [comment]
Test this please.

**Comment by burrimi** [comment]
test this please

**Pull request merged**